### ISSUE TITLE
Question about search space design and API

### ISSUE THREAD

**Issue by SimonBlanke** [SimonBlanke/Gradient-Free-Optimizers#12]
### This issue is a discussion about a question asked in another issue:

Hi @SimonBlanke sure, here are some examples on how these libraries support this.

## SciPy

SciPY defines the most common API for having the distribution as a function from which you can sample at will. This funcitonality is provided by the `rvs` (random variable sample) method, [example](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.rv_continuous.rvs.html).

[All the distributions in SciPy](https://docs.scipy.org/doc/scipy/reference/stats.html) follow this logic. For example a [loguniform](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.loguniform.html#scipy.stats.loguniform) and [uniform](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.uniform.html#scipy.stats.uniform).

This helps encoding the whole scan space without making a grid (which you do directly with the `np.arange`).

## Optuna

A [`trial` object](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.trial.Trial.html#optuna.trial.Trial) is passed as the argument of the objective function. The `trial` then can `suggest` values according to the distribution type and its bounds. The sampler then suggests from the available spaces that the trial presents to it. [Here](https://github.com/optuna/optuna/blob/9fe44349c05c9da38e4c8e84e9e048037e701862/optuna/samplers/_tpe/multivariate_parzen_estimator.py#L81) you can find the TPE sample method.

You see that it makes use of `rvs` methods, with the parameters of the distributions being learnt by the TPE.

## HyperOpt

Their codebase is a mess, but they usel `pyll` to define the spaces as distributions instead of arrays of examples, [here](https://github.com/hyperopt/hyperopt/blob/master/hyperopt/pyll/stochastic.py).

Their approach is similar to yours: they define the space before calling the objective function (as opposed to optuna). You can see that the workflow [here](https://github.com/hyperopt/hyperopt/blob/master/tutorial/02.MultipleParameterTutorial.ipynb) is very similar to your own.

## Scikit-optimize

They follow a similar approach again, where the space can be defined beforehand using distribution assignments. See [their Space API](https://scikit-optimize.github.io/stable/modules/classes.html#module-skopt.space.space). Again, notice how a dimension (i.e. a variable) has a `rvs` method.

For initial samplers, I'd recommend going through the `scikit-optimize` [`samplers` API](https://scikit-optimize.github.io/stable/modules/classes.html#module-skopt.sampler), where you can see different approaches to independent/warm-up samplings.

## Other examples

- [SMAC](https://automl.github.io/SMAC3/master/quickstart.html) also uses pre-defined variable bounds, within which it samples/operates.
- [Hyperas](https://github.com/maxpumperla/hyperas) as well, although it's not surprising as it is built in top of hyperopt.

## Final comments

All the cases above are Bayesian optimisation libraries, so these define the prior space over the variable space and the TPE, Gaussian Process, etc learn over these priors. But since they use the full prior (as oppose to a grid over it, as your methodology using `np.arange` imposes) you don't need to form these super huge grids.

The libraries I shown above, which I know better, are very focused on ML hyper-parameter tuning. I now have a different problem and would need grad-free optimisers (this is how I came across your library) and I am testing a few (I'm adapting my optuna, hyperopt, etc workflows to fit this new task) but I was very keen on trying the wealth of optimisers from your work.

_Originally posted by @romanovzky in https://github.com/SimonBlanke/Gradient-Free-Optimizers/issues/1#issuecomment-840696898_

**Comment by SimonBlanke** [comment]
After reading this overview thoroughly I still have some questions:

---

> SciPY defines the most common API for having the distribution as a function from which you can sample at will. This funcitonality is provided by the rvs (random variable sample) method

What would be the job of those distributions in GFO? Is this about how positions are initialized in the search space or how to sample from a continuous search space or something else? What would be the goal of doing it this other way? Why would this be "better" than the current implementation in GFO?

---

> All the cases above are Bayesian optimisation libraries, so these define the prior space over the variable space and the TPE, Gaussian Process, etc learn over these priors. But since they use the full prior (as oppose to a grid over it, as your methodology using np.arange imposes) you don't need to form these super huge grids.

This is about using continuous or discrete search spaces, right? When you implement smb-optimizers you still need to draw samples if the search space is continuous. Having a uniform grid of positions is (how I read it) the way to go, because you want to create an approximation of the score in entire search space without prioritizing a position at this stage of the algorithm (to explore the search space).

If you still want to reduce the computational load from doing predictions (from the surrogate model) on a huge grid you can still add a sampling algorithm. This is something I worked on and will be available soon.

**Comment by romanovzky** [comment]
So the way that I understand, using distributions from which you can sample would replace the current GFO workflow where you define the search spaces in a discritised manner even if you intend to have a continuous space.

Take the example from the main github page (the README):

```
search_space = {"x": np.arange(-10, 10, 0.1)}
```
These are 200 possibilities for the value of `x`, not the full interval. For example, it contains `9.8` and `9.9` but not `9.85`. Replacing this by `scipy.stats.uniform` I can sample how many points I want from the interval `[-10, 10]`:

```
from scipy.stats import uniform

points = uniform.rvs(-10, 10, 1000)
```

at this stage I see already two advantages of this approach:
- No need to invoke and store a long grid of possible values of intervals.
- Better coverage of the interval than a grid (this is akin to the customary discussion between grid and random searches)

The logic would then to be sample from these functions instead of from an `np.arange` of fixed grid points. The search is then conducted by knowing what are the bounds of these spaces (hence `scikit-opt` having the term `bounds` when referring to the space definition).

The way that I see it, this would also solve the issues you have with initialising the grid. In my use case, I have around 15 variables which I need to sample with considerable granularity from uniform intervals. GFO cannot start the process as it can't allocate the memory to construct the grid. In this sense GFO applicability is restricted to problems where it can draw a grid, which complexity grows geometrically with the number of variables. This will always happen for any logic that requires to build a N^D grid, where N is the number of points (200 in the above case) and D the number of dimensions.

`Scikit-opt` has a really nice `Samplers` API from which it draws initial points from these distributions (as I mentioned in the main post), effectively constructing an initial grid of random points without having to build it.

**Comment by SimonBlanke** [comment]
Lets see how GFOs works before continuing:
The search space is defined by creating a D-dimensional search space. Each dimension can have a different length than the others. 
For most optimization algorithms we already sample from a distribution (all hill climbing based algos). This can be seen in the "_move_climb"-method in the [hill climbing code](https://github.com/SimonBlanke/Gradient-Free-Optimizers/blob/master/gradient_free_optimizers/optimizers/local/hill_climbing_optimizer.py). So for most optimizers (except the sequence-model-based) we do not create a huge grid from the N-dimensional search space. 

So this point:
> No need to invoke and store a long grid of possible values of intervals.
is limited to the smbo algorithms.

The memory problem for smbo was already discussed in in issue #3 and if this is still a problem for you we should reopen the issue.

The memory-issue on those smbo is something that still bothers me, because you always have to take care not to create a search space that is to big. I will brainstorm again how to reimplement this part of the smbo-algorithms.

But just to be clear:
- Issue #1 is a separate problem that can be easily fixed without reimplementing the search space API.
- If the memory-problem of the smbos is fixed, I don't see any further problem of the search space API for now.

Do you agree with those points?

### Conclusion
I will take a closer look at the samplers in `Scikit-opt`, look into reimplementing the sampling of smbo-algorithms and reopen issue #3.

**Comment by romanovzky** [comment]
Probably I agree. I am user of gradient-free optimisers, but I am not a developer (I don't have a working knowledge of most of the algos beyond the high-level description).

Let me ask you a few more things:
- For the SMB case you keep the grid? In that case, wouldn't this bring in curse of dimensionality issues as the vertices of a D-dimensional hypercube are barely representative of its volume? In principle one can mitigate this with finer granularity of the grid, would GFO be able to work on "arbitrary" fine grids?
- Also for the SMB case, is it fundamentally necessary to be a grid (This is an educating questions, as I ask out of complete ignorance on how these are implemented)?

**Comment by SimonBlanke** [comment]
Hello @romanovzky,

About your first question:
Yes, currently we have to create a grid for those smbo, but we can currently sample from this grid after it was created. My goal is to do sampling without creating a (potentially large) grid beforehand. I don't understand the rest of your question. How would the vertices be barely representative? Vertices are often selected within the first few iterations of bayesian optimization, because the algorithm detects a large uncertainty at those positions (because they are often far away from the rest of the explored positions). This is the reason vertices are a part of the initialization in GFO. Most vertices are already evaluated beforehand by GFO, since bayes-opt will select them anyway. This saves time and computational resources.

---

About your second question:
Smbo works by collecting the positions and scores of each already evaluated position in the search space. A surrogate model is then trained on this information, where the positions are the features and the scores are the target. The prediction takes the positions from the grid to calculate the predicted scores in each position. From this we can calculate where the next best score (in the entire search space) might be. This next position and score is appended to the collection and we start fitting again. 
So we do not need a grid, but we need a limited number of positions.

---

The links you provided are very useful! The samplers from skopt cannot be used directly for GFO but I can implement those sampling algorithms myself and use them for the initialization and maybe the smbo sampling. 

Thank you very much for the help and information you provided. If you have any further questions, just let me know.
